include/smack/smack.h

```diff
@@ -135,32 +287,146 @@ void __SMACK_decls() {
+  D("function {:inline} $zext.i16.i32(p: i16) returns (i32) {(0bv16)++p}");
+  D("function {:inline} $zext.i32.i64(p: i32) returns (i64) {(0bv32)++p}");
+
+  D("function {:inline} $sext.i8.i64(p: i8) returns (i64) {if $sge.i8(p, 0bv8) then $zext.i8.i64(p) else (($neg.i64(1bv64))[64:8])++p}");
+  D("function {:inline} $sext.i8.i32(p: i8) returns (i32) {if $sge.i8(p, 0bv8) then $zext.i8.i32(p) else (($neg.i32(1bv32))[32:8])++p}");
+  D("function {:inline} $sext.i8.i16(p: i8) returns (i16) {if $sge.i8(p, 0bv8) then $zext.i8.i16(p) else $neg.i8(1bv8)++p}");
+  D("function {:inline} $sext.i16.i64(p: i16) returns (i64) {if $sge.i16(p, 0bv16) then $zext.i16.i64(p) else $neg.i64(1bv64)[64:16]++p}");
+  D("function {:inline} $sext.i16.i32(p: i16) returns (i32) {if $sge.i16(p, 0bv16) then $zext.i16.i32(p) else $neg.i16(1bv16)++p}");
+  D("function {:inline} $sext.i32.i64(p: i32) returns (i64) {if $sge.i32(p, 0bv32) then $zext.i32.i64(p) else $neg.i32(1bv32)++p}");
+
+  D("function {:inline} $p2i(p: ref) returns (size) {p}");
+  D("function {:inline} $i2p(p: size) returns (ref) {p}");
+  D("function {:inline} $p2b(p: ref) returns (bool) {p != $NULL}");
+#else
+  D("axiom $NULL == 0;");
   D("function {:inline} $pa(pointer: int, index: int, size: int) returns (int) {pointer + index * size}");
-  D("function {:inline} $trunc(p: int, size: int) returns (int) {p}");
+  D("function {:inline} $b2p(b: bool) returns (int) {if b then 1 else 0}");
   D("function {:inline} $p2i(p: int) returns (int) {p}");
   D("function {:inline} $i2p(p: int) returns (int) {p}");
   D("function {:inline} $p2b(p: int) returns (bool) {p != 0}");
-  D("function {:inline} $b2p(b: bool) returns (int) {if b then 1 else 0}");
+  D("function {:inline} $isExternal(p: int) returns (bool) { p < $GLOBALS_BOTTOM - 32768 }");
+  D("function {:inline} $trunc(p: int) returns (int) {p}");
+  D("function {:inline} $zext(p: int) returns (int) {p}");
+  D("function {:inline} $sext(p: int) returns (int) {p}");
+#endif
 
   // Memory debugging symbols
   D("type $mop;");
   D("procedure boogie_si_record_mop(m: $mop);");
   D("procedure boogie_si_record_bool(b: bool);");
+#ifdef BITVECTOR
+  D("procedure boogie_si_record_i8(i: i8);");
+  D("procedure boogie_si_record_i16(i: i16);");
+  D("procedure boogie_si_record_i32(i: i32);");
+  D("procedure boogie_si_record_i64(i: i64);");
+#endif
+  D("procedure boogie_si_record_ref(i: ref);");
   D("procedure boogie_si_record_int(i: int);");
   D("procedure boogie_si_record_float(f: float);");
   D("const $MOP: $mop;");
-  
-  D("const $GLOBALS_BOTTOM: int;");
-  D("function {:inline} $isExternal(p: int) returns (bool) { p < $GLOBALS_BOTTOM - 32768 }");
-  
+
+  D("const $GLOBALS_BOTTOM: ref;");
+
 #if MEMORY_MODEL_NO_REUSE_IMPLS
+#ifdef BITVECTOR
+  D("var $Alloc: [ref] bool;");
+  D("var $CurrAddr:ref;");
+
+  D("procedure $malloc(n: size) returns (p: ref)\n"
+    "modifies $CurrAddr, $Alloc;\n"
+    "{\n"
+    "  assume $sgt.ref($CurrAddr, $NULL);\n"
```

Comment:
Good point. Our unbounded-integer encoding of memory, which partitions heap and static storage by zero will not work with bit vectors. We will have to rethink that one.